### PR TITLE
Review and improve layout constraints

### LINKED ISSUE #4599: Wrapping output in center tag causes toolbar misalignment
As I've been updating our code to work with bokeh 0.12 I noticed that center aligning the notebook output using the html `center` tag causes misalignment of the toolbar. Here's a small code snippet that reproduces the issue:

``` python
from bokeh.io import notebook_div
from bokeh.models import HBox
from bokeh.plotting import figure
from IPython.display import display, HTML

output_notebook()

x = list(range(11))
y0 = x
y1 = [10 - i for i in x]
y2 = [abs(i - 5) for i in x]

# create a new plot
s1 = figure(width=300, plot_height=300, title=None)
s1.circle(x, y0, size=10, color="navy", alpha=0.5)

# create another one
s2 = figure(width=300, height=300, title=None)
s2.triangle(x, y1, size=10, color="firebrick", alpha=0.5)

# put all the plots in a VBox
p = HBox(children=[s1, s2])

# show the results
div = notebook_div(p)
display(HTML('<center>'+div+'</center>'))
```

image

Without the `<center>` tag it looks fine:

image

And I can get correct alignment by wrapping the output in a div like this `<div style="display:table; margin:0 auto;">`, so it isn't a massive priority.

This is using bokeh 0.12dev13 and Jupyter notebook 4.2.1.

### LINKED ISSUE #4972: Plot collapses on adding multiple labels above/below the plot if title='none' not specified in Plot()
Attached the images for reference. Without adding labels, one gets the normal plot (Fig 1). With one label added 'above', it works fine (Fig 2). When two labels are added though, whole plot collapses. (Fig 3)

Code:

```
from bokeh.plotting import show, output_file
from bokeh.models import Plot, Label, Circle, ColumnDataSource, DataRange1d, LinearAxis

output_file('gitex.html')

xdr = DataRange1d()
ydr = DataRange1d()

p = Plot(x_range=xdr, y_range=ydr, plot_width=300, plot_height=300, background_fill_color='#F0F0F0')

source = ColumnDataSource(dict(x=[1.0, 2.0, 3.0], y=[1.0, 2.0, 3.0]))

xaxis = LinearAxis()
p.add_layout(xaxis, 'below')

yaxis = LinearAxis()
p.add_layout(yaxis, 'left')

gly = Circle(x='x', y='y', size=10)
p.add_glyph(source, gly)

lab1 = Label(text='Label1', x=0, y=0, x_units='screen', y_units='screen')
p.add_layout(lab1, 'above')

lab2 = Label(text='Label2', x=0, y=0, x_units='screen', y_units='screen')
p.add_layout(lab2, 'above')

show(p)
```

Setting title=None in Plot() works as a quick fix. Mailing list reference for discussion on this with @birdsarah : https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/_dKphJePDwg
Figure1: 
figure1
Figure2: 
figure2
Figure3: 
figure3

### LINKED ISSUE #6578: No link to integration tests' report after PR #6542
This https://github.com/bokeh/bokeh/pull/6538#issuecomment-311080939 is the link I'm talking about.

### PR DESCRIPTION
**Pull request by mattpap**
This is an attempt to fix issues like #4972, where constraints are badly created, resulting in collapsed plots or "unsatisfiable constraint" error.

fixes #4972
fixes #4599 
fixes #6578

### CODE REVIEW

File: bokehjs/src/coffee/models/layouts/layout_dom.coffee [hunk 1]
```diff
@@ -261,24 +264,24 @@ export class LayoutDOM extends Model
...
+    @_height_minus_bottom = new Variable("#{@toString()}.height_minus_bottom")
     # Whitespace variables
-    @_whitespace_top = new Variable()
-    @_whitespace_bottom = new Variable()
-    @_whitespace_left = new Variable()
-    @_whitespace_right = new Variable()
+    @_whitespace_top = new Variable("#{@toString()}.whitespace_top")
+    @_whitespace_bottom = new Variable("#{@toString()}.whitespace_bottom")
+    @_whitespace_left = new Variable("#{@toString()}.whitespace_left")
+    @_whitespace_right = new Variable("#{@toString()}.whitespace_right")
```
Comment (bryevdv) [inline bokehjs/src/coffee/models/layouts/layout_dom.coffee]: These have alot of boilerplate can strings and var names to carefully match up. What about defining
```
_cv: (name) ->
    @['_'+name] = new Variable("{@toString()}.#{name})
```
then all these could be `@_cv('whitespace_right')`

**Comment by bryevdv** [comment]
@mattpap I had a quick look, and the code changes looked nice. I only had one suggestion to reduce some boilerplate. Report looks good, obviously all images are slightly diffed but I the changes seem either better or tolerable

File: bokehjs/src/coffee/models/layouts/layout_dom.coffee (hunk 1 above)
Comment (mattpap) [inline bokehjs/src/coffee/models/layouts/layout_dom.coffee]: Reducing boilerplate is important, but not like this. I will be taking care of this when rewriting this code in typescript.

Comment (bryevdv) [inline bokehjs/src/coffee/models/layouts/layout_dom.coffee]: OK if there is an even better method using TS then it can wait.

**Pull request merged**